### PR TITLE
KAFKA-8219: add doc changes for static membership release

### PR DESCRIPTION
**Pull request by abbccdda**
This is a doc-change only PR to add upgrade notes for static membership in 2.3

### CODE REVIEW

**Comment by abbccdda** [comment]
@guozhangwang @hachikuji @mjsax  Let me know if there is anything missing. Thank you!

**Comment by abbccdda** [comment]
Retest this please

**Comment by abbccdda** [comment]
Retest this please

**Comment by abbccdda** [comment]
work?

File: docs/streams/upgrade-guide.html [hunk 1]
```diff
@@ -191,6 +191,17 @@ <h3><a id="streams_api_changes_210" href="#streams_api_changes_210">Streams API
         details, see <a href="https://cwiki.apache.org/confluence/display/KAFKA/KIP-312%3A+Add+Overloaded+StreamsBuilder+Build+Method+to+Accept+java.util.Properties">KIP-312</a>
     </p>
 
+    <p>
+        We are introducing static membership towards KStream users. This feature reduces unnecessary rebalances during normal application upgrades or rolling bounces.
```
Comment (guozhangwang) [inline docs/streams/upgrade-guide.html]: `KStream` is not the standard term we recommend to call here. We should just use `Kafka Streams` or `Streams API`.

File: docs/streams/upgrade-guide.html [hunk 2]
```diff
@@ -191,6 +191,17 @@ <h3><a id="streams_api_changes_210" href="#streams_api_changes_210">Streams API
...
+        If you want to use static membership, first thing is to upgrade both broker cluster and stream apps to 2.3.0. This involves a necessity to change <code>inter.broker.protocol.version</code>
```
Comment (guozhangwang) [inline docs/streams/upgrade-guide.html]: `2.3.0` -> `2.3 or beyond`: we do not change web docs when release new bug fix versions e.g. 2.3.1 so we usually avoid indicating the specific bug-fix digit unless necessary.

Comment (guozhangwang) [reply docs/streams/upgrade-guide.html]: `This involves a necessity to change ...` -> `to 2.3 or beyond, and also make sure the upgraded brokers are using <code>inter.broker.protocol.version</code> of 2.3 or beyond as well.`

File: docs/streams/upgrade-guide.html [hunk 3]
```diff
@@ -191,6 +191,17 @@ <h3><a id="streams_api_changes_210" href="#streams_api_changes_210">Streams API
...
+        to 2.3 as well. Then you need to set the config <code>ConsumerConfig#GROUP_INSTANCE_ID_CONFIG</code> to a universally unique value for each stream instance.
+        If your broker is on older version, but you choose to set <code>ConsumerConfig#GROUP_INSTANCE_ID_CONFIG</code> on the client side, stream app will crash.
```
Comment (guozhangwang) [inline docs/streams/upgrade-guide.html]: `will crash` -> `will detect the broker version and then throws an UnsupportedException and shutdown`.

File: docs/upgrade.html [hunk 4]
```diff
@@ -61,6 +61,20 @@ <h4><a id="upgrade_2_2_0" href="#upgrade_2_2_0">Upgrading from 0.8.x, 0.9.x, 0.1
     </li>
 </ol>
 
+<h5><a id="upgrade_230_notable" href="#upgrade_230_notable">Notable changes in 2.3.0</a></h5>
+<ul>
+    <li>
+        We are introducing static membership towards consumer users. This feature reduces unnecessary rebalances during normal application upgrades or rolling bounces.
```
Comment (guozhangwang) [inline docs/upgrade.html]: Ditto as above.

File: docs/streams/upgrade-guide.html (hunk 1 above)
Comment (mjsax) [inline docs/streams/upgrade-guide.html]: We should use `KStream` only to refer to a stream (ie, `KStream` vs `KTable`)  and use `Kafka Streams` otherwise to avoid confusion.

File: docs/streams/upgrade-guide.html (hunk 2 above)
Comment (mjsax) [inline docs/streams/upgrade-guide.html]: `membership, [it is required] to upgrade your Kafka cluster (i.e., brokers) and your Kafka Streams application to 2.3.0.`

Comment (mjsax) [reply docs/streams/upgrade-guide.html]: Do we need to include `This involves a necessity to change <code>inter.broker.protocol.version</code> to 2.3 as well.` -- Maybe just link to the main upgrade section instead?

File: docs/streams/upgrade-guide.html [hunk 5]
```diff
@@ -191,6 +191,17 @@ <h3><a id="streams_api_changes_210" href="#streams_api_changes_210">Streams API
         details, see <a href="https://cwiki.apache.org/confluence/display/KAFKA/KIP-312%3A+Add+Overloaded+StreamsBuilder+Build+Method+to+Accept+java.util.Properties">KIP-312</a>
     </p>
 
+    <p>
+        We are introducing static membership towards KStream users. This feature reduces unnecessary rebalances during normal application upgrades or rolling bounces.
+        If you want to use static membership, first thing is to upgrade both broker cluster and stream apps to 2.3.0. This involves a necessity to change <code>inter.broker.protocol.version</code>
+        to 2.3 as well. Then you need to set the config <code>ConsumerConfig#GROUP_INSTANCE_ID_CONFIG</code> to a universally unique value for each stream instance.
```
Comment (mjsax) [inline docs/streams/upgrade-guide.html]: `Then you need to set the config` -> `You also need to set the config`

File: docs/streams/upgrade-guide.html (hunk 3 above)
Comment (mjsax) [inline docs/streams/upgrade-guide.html]: Do we need this?

File: docs/streams/upgrade-guide.html [hunk 6]
```diff
@@ -191,6 +191,17 @@ <h3><a id="streams_api_changes_210" href="#streams_api_changes_210">Streams API
...
+        If your broker is on older version, but you choose to set <code>ConsumerConfig#GROUP_INSTANCE_ID_CONFIG</code> on the client side, stream app will crash.
+        If you accidentally configure duplicate ids for different stream instances, a fencing mechanism will inform your duplicate client to shutdown by
```
Comment (mjsax) [inline docs/streams/upgrade-guide.html]: Do we need this?

**Review by mjsax — COMMENTED** [review]
My comments on Streams upgrade docs also apply to the main docs -- just did not add them twice.

For the things, that I marked as "do we need this", it might be better to add a proper section about static group membership and explain how it works. And we could link to this section. In future versions, people won't read older release notes and thus this information get buried -- having a proper section about static group membership avoid this issue.

File: docs/streams/upgrade-guide.html (hunk 2 above)
Comment (abbccdda) [reply docs/streams/upgrade-guide.html]: I guess declaring the requirement here is important to make sure user won't choose to only upgrade one broker for coordinator as a shortcut.

File: docs/streams/upgrade-guide.html (hunk 3 above)
Comment (abbccdda) [reply docs/streams/upgrade-guide.html]: We need it for user to trouble shooting when they set group.instance.id on client where broker doesn't support.

File: docs/streams/upgrade-guide.html (hunk 6 above)
Comment (abbccdda) [reply docs/streams/upgrade-guide.html]: We need to document the expected mechanism for dealing with duplicate instance id case.

**Comment by abbccdda** [comment]
Working on a central page for static membership.

File: docs/streams/upgrade-guide.html (hunk 2 above)
Comment (mjsax) [reply docs/streams/upgrade-guide.html]: Yes, but this information is in the broker upgrade section already -- it seems better to not scatter information, but link to the corresponding upgrade sections? Or maybe I am missing your point?

File: docs/streams/upgrade-guide.html (hunk 3 above)
Comment (mjsax) [reply docs/streams/upgrade-guide.html]: Yes, but as mentioned below, I would prefer to have this information is a proper "static group membership" sections. Having it in the upgrade guide does not seem to be the best place.

File: docs/design.html [hunk 7]
```diff
@@ -216,6 +216,26 @@ <h4><a id="design_offlineload" href="#design_offlineload">Offline Data Load</a><
...
     management, and tasks which fail can restart without danger of duplicate data&mdash;they simply restart from their original position.
 
+    <h4><a id="static_membership" href="#static_membership">Static Membership</a></h4>
+    Static membership aims to improve the availability of stream applications, consumer groups and other applications built on top of the group rebalance protocol.
+    The rebalance protocol relies on the group coordinator to allocate entity ids to group members. These generated ids are ephemeral and will change when members restart and rejoin.
+    For consumer based apps, this "dynamic membership" can cause a large percentage of tasks re-assigned to different instances during administrative operations
+    such as code deploys, configuration updates and periodic restarts. For large state applications, shuffled tasks need a long time to recover their local states before processing
+    and cause applications to be partially or entirely unavailable. Motivated by this observation, new Kafka’s group management protocol allows group members to provide persistent entity ids.
```
Comment (mjsax) [inline docs/design.html]: nit: remove `new`

File: docs/design.html [hunk 8]
```diff
@@ -216,6 +216,26 @@ <h4><a id="design_offlineload" href="#design_offlineload">Offline Data Load</a><
...
+    Group membership remains unchanged based on those ids, thus no rebalance will be triggered.
+    <p>
+    If you want to use static membership,
+    <ul>
+        <li>Upgrade both broker cluster and stream apps to 2.3 or beyond, and also make sure the upgraded brokers are using <code>inter.broker.protocol.version</code>
```
Comment (mjsax) [inline docs/design.html]: nit remove `stream`

File: docs/design.html [hunk 9]
```diff
@@ -216,6 +216,26 @@ <h4><a id="design_offlineload" href="#design_offlineload">Offline Data Load</a><
...
+            of 2.3 or beyond as well.</li>
+        <li>Set the config <code>ConsumerConfig#GROUP_INSTANCE_ID_CONFIG</code> to a unique value for each consumer instance under one group.</li>
```
Comment (mjsax) [inline docs/design.html:231]: Maybe add one bullet point for Streams:
`<li>For Kafka Streams applications, it is sufficient to set a unique <code>ConsumerConfig#GROUP_INSTANCE_ID_CONFIG</code> per `KafkaStreams` instance, independent of the number of used threads for an instance.</li>`

Side question: what happens if one set's the config for a consumer with no group.id -- this case could easily happen for Kafka Streams restore and global consumer?

File: docs/design.html [hunk 10]
```diff
@@ -216,6 +216,26 @@ <h4><a id="design_offlineload" href="#design_offlineload">Offline Data Load</a><
...
+    </ul>
+    If your broker is on older version, but you choose to set <code>ConsumerConfig#GROUP_INSTANCE_ID_CONFIG</code> on the client side, stream app will detect the broker version and
```
Comment (mjsax) [inline docs/design.html]: `is on [an] older version [than 2.3],`

`stream app` -> `the application`

File: docs/design.html [hunk 11]
```diff
@@ -216,6 +216,26 @@ <h4><a id="design_offlineload" href="#design_offlineload">Offline Data Load</a><
...
+    then throws an UnsupportedException and shutdown. If you accidentally configure duplicate ids for different stream instances,
```
Comment (mjsax) [inline docs/design.html]: remove `and shotdown` (only true for Kafka Streams, but not for plain consumer)

`stream instances` -- the docs should be written in a non-streams specific way here

**Review by mjsax — COMMENTED** [review]
Thanks for the update! I like the new section a lot!

File: docs/design.html (hunk 9 above)
Comment (abbccdda) [reply docs/design.html:231]: @mjsax Static membership only affects group based consumer app. Standalone consumers are not affected (I.E. we won't assign them `group.instance.id`

Comment (mjsax) [reply docs/design.html:231]: Let me rephrase: What happens if a user configures no `group.id` but does configure `group.instance.id`. Do the consumer crash? Or is the `group.instance.id` config just ignored?

Comment (abbccdda) [reply docs/design.html:231]: My understanding is that these two configs should not affect each other. But will try to add a unit test to check.

**Comment by abbccdda** [comment]
@guozhangwang Could you take another look?

**Comment by guozhangwang** [comment]
Cherry-picked to 2.3 as well.
